### PR TITLE
fix the bug in pipeline data parallelism

### PR DESCRIPTION
**Pull request by sandyhouse**
### PR types

Bug fixes
### PR changes

Others
### Describe

Fix the bug in pipeline data parallelism resulting in data parallelism cannot work correctly.

### CODE REVIEW

File: python/paddle/distributed/fleet/meta_optimizers/pipeline_optimizer.py [hunk 1]
```diff
@@ -41,35 +41,36 @@ def update_startup_program(self,
...
-        self._init_communicator(self.startup_program, current_endpoint,
-                                pipeline_endpoints, pipeline_rank, 0,
-                                self.wait_port)
+        endpoints = self.role_maker._get_trainer_endpoints()
+        current_endpoint = endpoints[rank]
+        node_num = _get_node_num(endpoints)
+        assert nranks % node_num == 0
+
+        # Create ring 0 for all gpus in the same pipeline
+        if inner_parallelism > 1:
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/pipeline_optimizer.py:52]: how if inner_parallelism <= 1

File: python/paddle/distributed/fleet/meta_optimizers/pipeline_optimizer.py [hunk 2]
```diff
@@ -41,35 +41,36 @@ def update_startup_program(self,
...
+            pipeline_endpoints = endpoints[start_index:start_index +
+                                           inner_parallelism]
+            self._init_communicator(self.startup_program, current_endpoint,
+                                    pipeline_endpoints, pipeline_rank, 0,
+                                    self.wait_port)
 
         pipeline_num = len(endpoints) // inner_parallelism
         if pipeline_num == 1: return
-        # Create rings for gpus with the same gpu id
+        # Create rings for gpus with the same pipeline id for data parallel
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/pipeline_optimizer.py:64]: 这里面注释可以写清楚一点，给一个例子来描述一下这段代码逻辑
0 1 2 3 4 5 6 7
当inner_parallelism为4时，
(0 1 2 3) (4 5 6 7)为inner流水线并行，用ring 0
(0 4) (1 5) (2 6) (3 7)各自组环，为数据并行

File: python/paddle/distributed/fleet/meta_optimizers/pipeline_optimizer.py [hunk 3]
```diff
@@ -108,8 +109,10 @@ def _init_communicator(self, program, current_endpoint, endpoints, rank,
 
     def _broadcast_params(self, ring_id):
         block = self.startup_program.global_block()
-        for param in block.iter_parameters():
-            if param.is_distributed:
+        for var_name in block.vars:
+            if "nccl_id" in var_name: continue
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/pipeline_optimizer.py:113]: maybe can use var().type() != core.VarDesc.VarType.RAW

File: python/paddle/distributed/fleet/meta_optimizers/pipeline_optimizer.py [hunk 4]
```diff
@@ -41,35 +41,36 @@ def update_startup_program(self,
...
+            pipeline_rank = rank % inner_parallelism
+            pipeline_id = rank // inner_parallelism
+            start_index = pipeline_id * inner_parallelism
+            pipeline_endpoints = endpoints[start_index:start_index +
+                                           inner_parallelism]
+            self._init_communicator(self.startup_program, current_endpoint,
+                                    pipeline_endpoints, pipeline_rank, 0,
+                                    self.wait_port)
 
         pipeline_num = len(endpoints) // inner_parallelism
```
Comment (wangxicoding) [inline python/paddle/distributed/fleet/meta_optimizers/pipeline_optimizer.py:62]: 这个名字有点晦涩，是不是叫dp_nranks更好些。

File: python/paddle/fluid/optimizer.py [hunk 5]
```diff
@@ -3840,45 +3843,65 @@ def _split_program(self, main_program, devices):
                     op_desc = op.desc
                     ap_op = program["program"].block(0).desc.append_op()
                     ap_op.copy_from(op_desc)
-                    ap_op._set_attr(self._op_device_key, "")
-            elif op.type == "create_py_reader" or op.type == "read":
+                    # ap_op._set_attr(self._op_device_key, "")
```
Comment (wangxicoding) [inline python/paddle/fluid/optimizer.py:3846]: remove?

File: python/paddle/distributed/fleet/meta_optimizers/pipeline_optimizer.py (hunk 1 above)
Comment (sandyhouse) [reply python/paddle/distributed/fleet/meta_optimizers/pipeline_optimizer.py:52]: If inner_parallelism <=1, which means only one device is used in a pipeline, we have no need to create a ring for pipeline, and only data parallel is considered.

File: python/paddle/distributed/fleet/meta_optimizers/pipeline_optimizer.py (hunk 2 above)
Comment (sandyhouse) [reply python/paddle/distributed/fleet/meta_optimizers/pipeline_optimizer.py:64]: I will add a comment in a new pr.

File: python/paddle/fluid/optimizer.py (hunk 5 above)
Comment (sandyhouse) [reply python/paddle/fluid/optimizer.py:3846]: I will remove it in a new pr.

**Review by wangxicoding — APPROVED** [review]
LGTM

**Review by fuyinno4 — APPROVED** [review]

**Review by chalsliu — APPROVED** [review]

**Pull request merged**